vllm/attention/ops/blocksparse_attention/blocksparse_attention_kernel.py

```diff
@@ -3,6 +3,407 @@
+            k = tl.load(K_cache + off_k,
+                        mask=(start_n + offs_n[None, :]) < context_len,
+                        other=0.0)  # [D,N]
+            v = tl.load(V_cache + off_v,
+                        mask=(start_n + offs_n[:, None]) < context_len,
+                        other=0.0)  # [N,D]
+            acc, l_i, m_i = _context_fwd_kernel_inner(
+                acc,
+                l_i,
+                m_i,
+                q,
+                Q,
+                k,
+                v,
+                sm_scale,
+                sparse_block_size,
+                sparse_block_size,
+                # last key block in the key cache
+                k_block_id == context_end_block_id,
+                offs_m,
+                context_len,
+                start_n,
+                offs_n)
+
+        # compute query against itself (with causal mask)
+        if k_block_id >= query_start_block_id:
+            start_n = k_block_id * sparse_block_size
+            next_k_block_start_location = (k_block_id + 1) * sparse_block_size
+            key_offset = 0 if k_block_id == query_start_block_id \
+                else start_n - context_len
+            offs_n = key_offset + tl.arange(0, sparse_block_size)
+            # -- compute qk ----
+            k = tl.load(
+                K + offs_n[None, :] * stride_kbs + offs_d[:, None] * stride_kd,
+                # only load current query block, and also could not exceed query
+                # length
+                mask=(offs_n[None, :] < query_len) &
+                (offs_n[None, :] + context_len < next_k_block_start_location),
+                other=0.0)
+            v = tl.load(
+                V + offs_n[:, None] * stride_vbs + offs_d[None, :] * stride_vd,
+                # only load current query block, and also could not exceed query
+                # length
+                mask=(offs_n[:, None] < query_len) &
+                (offs_n[:, None] + context_len < next_k_block_start_location),
+                other=0.0)
+            acc, l_i, m_i = _context_fwd_kernel_inner(
+                acc,
+                l_i,
+                m_i,
+                q,
+                Q,
+                k,
+                v,
+                sm_scale,
+                sparse_block_size,
+                sparse_block_size,
+                # LAST_K_BLOCK
+                k_block_id == block_id_for_current_query_block,
+                offs_m,
+                context_len,
+                context_len,
+                offs_n)
+    # flash-attn 2
+    # m_i += tl.math.log2(l_i)
+    acc = acc / l_i[:, None]
+    # write output
+    tl.store(Out + offs_m[:, None] * stride_obs + offs_d[None, :] * stride_od,
+             acc,
+             mask=(offs_m[:, None] < query_len) &
+             (offs_m[:, None] + context_len < next_q_block_start_location))
+    return
+
+
+@torch.inference_mode()
+def context_blocksparse_flash_attn_varlen_fwd(
```

Comment:
If math equation is added, people can improve and replace it easier in the future.